File: src/UnwindMC/Generation/Ast/Ast.fs
module Ast

open System.Collections.Generic

type Operator =
    | Equal
    | NotEqual
    | Less
    | LessOrEqual
    | Greater
    | GreaterOrEqual

    | Not
    | And
    | Or
    | Xor
    | ShiftLeft
    | ShiftRight

    | Negate
    | Add
    | Subtract
    | Multiply
    | Divide
    | Modulo

type Var = Var of string

type Expression =
    | Binary of Operator * Expression * Expression
    | Dereference of Expression
    | Unary of Operator * Expression
    | Value of int
    | VarRef of Var

type Statement =
    | Assignment of Var * Expression
    | Break
    | Continue
    | DoWhile of IReadOnlyList<Statement> * Expression
    | For of Expression * IReadOnlyList<Statement> * IReadOnlyList<Statement>
    | FunctionCall of Expression
    | IfThenElse of Expression * IReadOnlyList<Statement> * IReadOnlyList<Statement>
    | Return of Var option
    | While of Expression * IReadOnlyList<Statement>

File: src/ast.rs
#[derive(Debug, PartialEq, Clone, Copy)]
pub enum Operator {
    Equal,
    NotEqual,
    Less,
    LessOrEqual,
    Greater,
    GreaterOrEqual,

    Not,
    And,
    Or,
    Xor,
    ShiftLeft,
    ShiftRight,

    Negate,
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Var {
    Var(String),
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expression {
    Binary(Operator, Box<Expression>, Box<Expression>),
    Dereference(Box<Expression>),
    Unary(Operator, Box<Expression>),
    Value(i32),
    VarRef(Var),
}

#[derive(Debug, PartialEq, Clone)]
pub enum Statement {
    Assignment(Var, Expression),
    Break,
    Continue,
    DoWhile(Vec<Statement>, Expression),
    For(Expression, Vec<Statement>, Vec<Statement>),
    FunctionCall(Expression),
    IfThenElse(Expression, Vec<Statement>, Vec<Statement>),
    Return(Option<Var>),
    While(Expression, Vec<Statement>),
}
